### PR TITLE
teams discovery API

### PR DESCRIPTION
**Pull request by neha-Gupta1**
Teams discovery API
- fetch all groups from endpoint and filter teams out of it.
- fetch team with ID. 

Permission added- 
- Application- Team.ReadBasic.All
- Group.Read.All

#### Does this PR need a docs update or release note?
- [ ] :clock1: Yes, but in a later PR

#### Type of change

- [ ] :sunflower: Feature

#### Issue(s)

* https://github.com/alcionai/corso/issues/3836

#### Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

**Comment by neha-Gupta1** [comment]
Testcases and few error cases still to be handled.

File: src/pkg/services/m365/api/teams.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			if s == teamService {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// GetID retrieves team by groupID/teamID.
+func (c Groups) GetByID(
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:117]: On the topic of supporting WebURLs as a teams identifier, we'll need an api for looking up teams by either id or weburl.  That change should be saved for a separate PR, just wanted to make the need clear.

File: src/pkg/services/m365/api/teams.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	ctx context.Context,
+	identifier string,
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:119]: Can this fetch a team using a team ID, or does it need a group ID that is different from the team id?

File: src/pkg/services/m365/api/teams.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := service.Client().Groups().ByGroupId(identifier).Get(ctx, nil)
+	if err != nil {
+		return nil, graph.Wrap(ctx, err, "getting group by ID")
+	}
+
+	if err != nil {
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:131]: duplicate conditionals

File: src/pkg/services/m365/api/teams.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		err := graph.Wrap(ctx, err, "getting teams by id")
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:132]: You're not necessarily getting teams here (unless you change the func, as suggested above).
```suggestion
		err := graph.Wrap(ctx, err, "getting groups by id")
```

File: src/pkg/services/m365/api/teams.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+		// TODO: check if its applicable here
+		if graph.IsErrItemNotFound(err) {
+			err = clues.Stack(graph.ErrResourceOwnerNotFound, err)
+		}
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:137]: Let's skip the addition of any code that we're not certain will be used.

File: src/pkg/services/m365/api/teams.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+		return nil, err
+	}
+
+	return resp, err
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:142]: Returns should be one of:
```go
return resp, nil
return nil, resp
return resp, graph.Stack(ctx, err).OrNil()
```

File: src/pkg/services/m365/api/teams.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+// ValidateGroup ensures the item is a Groupable, and contains the necessary
+// identifiers that we handle with all groups.
+func ValidateGroup(item models.Groupable) error {
+	if item.GetId() == nil {
+		return clues.New("missing ID")
+	}
+
+	if item.GetDisplayName() == nil {
+		return clues.New("missing principalName")
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:157]:
```suggestion
		return clues.New("missing display name")
```

File: src/pkg/services/m365/m365.go
Comment (ryanfkeepers) [inline src/pkg/services/m365/m365.go:1]: Please move this file to a separate PR.  It'll keep each review smaller, and give more space to focus on the addition of testing as necessary.

File: src/pkg/services/m365/api/team_test.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+type TeamsUnitSuite struct {
+	tester.Suite
+}
+
+func TestTeamsUnitSuite(t *testing.T) {
+	suite.Run(t, &TeamsUnitSuite{Suite: tester.NewUnitSuite(t)})
+}
+
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/team_test.go:22]: this isn't getting used.

File: src/pkg/services/m365/api/team_test.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	t := suite.T()
+
+	ctx, flush := tester.NewContext(t)
+	defer flush()
+
+	teams, err := suite.its.ac.
+		Groups().
+		GetAll(ctx, true, fault.New(true))
+	require.NoError(t, err)
+	require.NotZero(t, len(teams), "must have at least one team")
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/team_test.go:50]: This is a weak test that doesn't validate much for us beyond our ability to call the func in the first place.  I'd really like to see the default test team set up in the tester package as part of this PR, and for the tests here to ensure that team can be found when we call `GetAll`, and also when we call `GetByID`.

File: src/pkg/services/m365/api/team_test.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+func TestTeamsIntgSuite(t *testing.T) {
+	suite.Run(t, &TeamsIntgSuite{
+		Suite: tester.NewIntegrationSuite(
+			t,
+			[][]string{tconfig.M365AcctCredEnvs}),
+	})
+}
+
+func (suite *TeamsIntgSuite) SetupSuite() {
+	suite.its = newIntegrationTesterSetup(suite.T())
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/team_test.go:37]: integrationTesterSetup should have teams (and groups?) details added to it, so that the data is available for all tests as needed.

**Review by ryanfkeepers — CHANGES_REQUESTED** [review]
Generally looks good, though some minor refactoring is wanted.  The two blocking changes are:

1. the PR should be split into two PRs.  One for the api file.  One for api/teams.go, one for the addition of teams to services/m365.
2. Testing setup needs to be complete and the api tests for groups needs to be expanded.

File: src/pkg/services/m365/api/teams.go (hunk 7 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: Done

File: src/pkg/services/m365/api/teams.go (hunk 2 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: With my understanding/investigation - value of groupID created for the team is same as TeamID.

File: src/pkg/services/m365/api/teams.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+// ---------------------------------------------------------------------------
+
+func (c Client) Groups() Groups {
+	return Groups{c}
+}
+
+// On creation of each Microsoft Teams a corrsponding group gets created from them.
+// Most of the information like events, drive and mail info will be fetched directly
+// from groups. So we pull in group and process only the once which are associated with
+// a team for further proccessing of teams.
```
[the remark this answers is not in the document]
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: I have updated the comments.

File: src/pkg/services/m365/api/teams.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+	"github.com/alcionai/clues"
+	"github.com/alcionai/corso/src/internal/common/str"
+	"github.com/alcionai/corso/src/internal/common/tform"
+	"github.com/alcionai/corso/src/internal/m365/graph"
+	"github.com/alcionai/corso/src/pkg/fault"
+	msgraphgocore "github.com/microsoftgraph/msgraph-sdk-go-core"
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+)
+
+const teamService = "Team"
```
[the remark this answers is not in the document]
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: `Team` is the value endpoint provides. So will keep it same. Updated the const name.

File: src/pkg/services/m365/m365.go
Comment (neha-Gupta1) [reply src/pkg/services/m365/m365.go:1]: Moved to - https://github.com/alcionai/corso/pull/3864

File: src/pkg/services/m365/api/teams.go [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+		return nil, err
+	}
+
+	return getGroups(ctx, filterTeams, errs, service)
+}
+
+// GetAll retrieves all groups.
+func getGroups(
+	ctx context.Context,
+	filterTeams bool,
```
[the remark this answers is not in the document]
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: Exposed two funcs now- 
- func (c Groups) GetAllTeams() - gets all the groups which have a corresponding team with them.
- func (c Groups) GetAll() - gets all the groups in org

File: src/pkg/services/m365/api/teams.go (hunk 1 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: Will handle it in -- https://github.com/alcionai/corso/pull/3864

File: src/pkg/services/m365/m365.go
Comment (ryanfkeepers) [reply src/pkg/services/m365/m365.go:1]: thank you!

File: src/pkg/services/m365/api/teams.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/pkg/services/m365/api/teams.go]: Sounds good.  In that case, we'll want this func to accept any variation of identifier we use for the team.  For example: sites can provide either the site id or the webURL, users can provide either user id or principal name.  Given that not all groups are teams, we might want to make a `GetTeamByID` func instead.

Though, now that I think about it, maybe betwen this and the GetAllTeams suggestion, this is a good time to swap this client from `Groups` to `Teams`, and have the teams api call groups under the hood?  It'll probably be clearer for us to make calls like `ac.Teams().GetByID()` and `ac.Teams().GetGroupByID()` than it would be to have lots of `ac.Groups().GetFooByID()` variations.

File: src/pkg/services/m365/api/teams.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/pkg/services/m365/api/teams.go]: Ah, good catch, thanks.

File: src/pkg/services/m365/api/teams.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/pkg/services/m365/api/teams.go]: See my other comment: I think we'll want two clients.  One for teams, and one for groups, and the teams client can call groups api under the hood.  For this PR, since we don't have a need for generalized groups yet, can you just shift the api client naming to teams?  Thanks for handling all the-back-and-forth.

File: src/pkg/services/m365/api/teams.go (hunk 1 above)
Comment (ryanfkeepers) [reply src/pkg/services/m365/api/teams.go]: hmm, I think this PR is the correct place to handle it.  Or, at least, the api layer should own that handling, not the discovery layer.  This change can move to a follow-up PR, though we'll need a TODO for it.

File: src/pkg/services/m365/api/teams.go (hunk 2 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: Done.

File: src/pkg/services/m365/api/teams.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: As of now I have added the Teams client with uses group api under the hood.

File: src/pkg/services/m365/api/team_test.go (hunk 9 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/team_test.go]: Done

File: src/pkg/services/m365/api/team_test.go (hunk 8 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/team_test.go]: Updated it.

File: src/pkg/services/m365/api/team_test.go (hunk 10 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/team_test.go]: Added helper functions.

File: src/pkg/services/m365/api/groups.go [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+func (c Teams) GetAll(
+	ctx context.Context,
+	errs *fault.Bus,
+) ([]models.Groupable, error) {
+	service, err := c.Service()
+	if err != nil {
+		return nil, err
+	}
+
+	return getGroups(ctx, true, errs, service)
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/groups.go:49]: I'm  playing around with graph api and it looks like the call to `/teams` works just fine as an enumerator, instead of calling `/groups` and filtering down to teams.  

On that note, this file is still conflicted about its identity.  It's called `Groups`, but has a client named `Teams`.  The teams client has a `GetAll` query that calls a `Groups` endpoint.  This all compiles and works out fine, but it's going to be a mess for us to maintain in the long term.  Lets keep it simple for now and have something like this:

```
// in /api/teams.go

func (c Client) Teams() Teams

func (c Teams) GetAll(ctx, errs)
    // should call `/teams`
    
func (c Teams) GetByID(ctx, id, errs)
    // should call `/teams/<id>`
    
func ValidateTeam(...)
```

Simple as that.  Ditch anything groups related until we have clear reason to need it.

Comment (neha-Gupta1) [reply src/pkg/services/m365/api/groups.go:49]: Yes. That was my first try. But I could not find the GET method for teams in the sdk. Here is the ticket for same - https://github.com/microsoftgraph/msgraph-sdk-go/issues/539. 
Till then I am try to use the raw API.

Comment (neha-Gupta1) [reply src/pkg/services/m365/api/groups.go]: With latest msgraph SDK it works fine. Updated the code.

Comment (neha-Gupta1) [reply src/pkg/services/m365/api/groups.go]: Will require https://github.com/alcionai/corso/pull/3879 to be merged.

File: src/internal/tester/tconfig/config.go [hunk 15]
```diff
@@ -38,6 +40,8 @@ const (
 const (
 	EnvCorsoM365TestSiteID           = "CORSO_M365_TEST_SITE_ID"
 	EnvCorsoM365TestSiteURL          = "CORSO_M365_TEST_SITE_URL"
+	EnvCorsoM365TestTeamID           = "CORSO_M365_TEST_TEAM_ID"
+	EnvCorsoM365TestGroupID          = "CORSO_M365_TEST_Group_ID"
```
Comment (ryanfkeepers) [inline src/internal/tester/tconfig/config.go:44]:
```suggestion
	EnvCorsoM365TestGroupID          = "CORSO_M365_TEST_GROUP_ID"
```

File: src/internal/tester/tconfig/config.go [hunk 16]
```diff
@@ -148,6 +152,18 @@ func ReadTestConfig() (map[string]string, error) {
 		os.Getenv(EnvCorsoM365TestSiteID),
 		vpr.GetString(TestCfgSiteID),
 		"10rqc2.sharepoint.com,4892edf5-2ebf-46be-a6e5-a40b2cbf1c1a,38ab6d06-fc82-4417-af93-22d8733c22be")
+	fallbackTo(
+		testEnv,
+		TestCfgTeamID,
+		os.Getenv(EnvCorsoM365TestTeamID),
+		vpr.GetString(TestCfgTeamID),
+		"d288d6bc-4595-4ff5-87a1-6e7fd750aa42")
```
Comment (ryanfkeepers) [inline src/internal/tester/tconfig/config.go:160]: any particular reason to have a separate group and team?

Either way, for the team, can we please use `6f24b40d-b13d-4752-980f-f5fb9fba7aa0` (CorsoCITeam).

File: src/internal/tester/tconfig/protected_resources.go [hunk 17]
```diff
@@ -209,3 +209,29 @@ func UnlicensedM365UserID(t *testing.T) string {
...
+	cfg, err := ReadTestConfig()
+	require.NoError(t, err, "retrieving m365 team id from test configuration: %+v", clues.ToCore(err))
+
+	return strings.ToLower(cfg[TestCfgTeamID])
+}
+
+// Groups
+
+// M365GroupID returns a groupID string representing the m365GroupID described
+// by either the env var CORSO_M365_TEST_Group_ID, the corso_test.toml config
```
Comment (ryanfkeepers) [inline src/internal/tester/tconfig/protected_resources.go:229]:
```suggestion
// by either the env var CORSO_M365_TEST_GROUP_ID, the corso_test.toml config
```

File: src/pkg/services/m365/api/teams.go [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+// The group acts as the protected resource, and all teams data like events,
+// drive and mail messages are owned by that group.
+
+// Teams is an interface-compliant provider of the client.
+type Teams struct {
+	Client
+}
+
+// GetAllTeams retrieves all groups.
+func (c Teams) GetAll(
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:40]: Apologies again on the design thrashing, but per our sync conversations, can you please change this package, file, and client backup to `Groups` instead of `Teams`?  That seems like it'll be a more appropriate starting point for the work we plan to implement.  If we move from there to needing a teams specific api, we can introduce another client for that.  Once the client naming is squared away, this is good to go.
```suggestion
func (c Groups) GetAll(
```

File: src/pkg/services/m365/api/teams.go [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+		return nil, err
+	}
+
+	return getGroups(ctx, true, errs, service)
+}
+
+// GetAll retrieves all groups.
+func getGroups(
+	ctx context.Context,
+	getOnlyTeams bool,
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/teams.go:55]: As mentioned in the other comment, let's ditch this for now and get all groups instead.

That said, not all groups have access to conversations (and, I assume, groups without access to conversations also don't have access to a calendar).  Instead of packing that filter into the pagination, how about a separate func that can accept a `[]models.Groupable` as a parameter and filter it down to only specific groups, such as only those which can access conversations?  This change can be made in a follow-up PR.

File: src/pkg/services/m365/api/teams.go (hunk 18 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: done

File: src/pkg/services/m365/api/teams.go (hunk 19 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/teams.go]: Done. Currently we have only one such func implemented for teams. But it can be extended in future.

File: src/pkg/services/m365/api/groups.go [hunk 20]
```diff
@@ -46,13 +46,26 @@ func (c Teams) GetAll(
...
+	}
+
+	return getGroups(ctx, FetchOnlyTeams, errs, service)
 }
 
 // GetAll retrieves all groups.
 func getGroups(
 	ctx context.Context,
-	getOnlyTeams bool,
+	filterGroupsData func(ctx context.Context, g models.Groupable) bool,
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/groups.go:68]: Hmm, this isn't my preferred implementation.  Ad-hoc funcs for filtering are genrally a more complicated implementation of a standalone transformer func.  I think a more simple implementation is to split this out to the following funcs.  These are each more simple to grok, more simple to test, and can be exported if we need their behavior elsewhere.

```go
// gets all groups, does no filtering

func (c Client) getGroups(ctx, errs) ([]models.Groupable, error) {...}

// callers can pass the results of getGroups() here.

func removeTeams([]models.Groupable) []models.Groupable {...}

func removeNonTeams([]models.Groupable) []models.Groupable {...}
```

**Review by ryanfkeepers — APPROVED** [review]
one last design nit.  approving to unblock.

File: src/pkg/services/m365/api/groups.go (hunk 20 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/groups.go:68]: Got it. Something similar to what I implemented initially.

File: src/pkg/services/m365/api/groups.go [hunk 21]
```diff
@@ -111,25 +113,28 @@ func getGroups(
...
-				log.Debug("could not be converted to string value: ", ResourceProvisioningOptions)
-				return false
-			}
-
-			if s == teamsAdditionalDataLabel {
-				return true
+	var teams []models.Groupable
+
+	for _, g := range groups {
+		if g.GetAdditionalData()[ResourceProvisioningOptions] != nil {
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/groups.go:121]:
```suggestion
		if g.GetAdditionalData()[ResourceProvisioningOptions] != nil {
		  continue
		}
		// the rest
```

File: src/pkg/services/m365/api/groups.go [hunk 22]
```diff
@@ -111,25 +113,28 @@ func getGroups(
...
+			for _, v := range val {
+				s, err := str.AnyToString(v)
+				if err != nil {
+					log.Debug("could not be converted to string value: ", ResourceProvisioningOptions)
+					continue
+				}
+
+				if s == teamsAdditionalDataLabel {
+					teams = append(teams, g)
+				}
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/groups.go:132]: I forget, do we already have this in multiple spots?  If so, can it be moved into a helper func?

File: src/pkg/services/m365/api/groups_test.go [hunk 23]
```diff
@@ -116,15 +116,25 @@ func (suite *GroupsIntgSuite) TestGetAllTeams() {
 	ctx, flush := tester.NewContext(t)
 	defer flush()
 
-	groups, err := suite.its.ac.
+	teams, err := suite.its.ac.
 		Groups().
 		GetTeams(ctx, fault.New(true))
 	require.NoError(t, err)
+	require.NotZero(t, len(teams), "must have at least one group")
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/groups_test.go:123]:
```suggestion
	require.NotZero(t, len(teams), "must have at least one team")
```

File: src/pkg/services/m365/api/groups_test.go [hunk 24]
```diff
@@ -116,15 +116,25 @@ func (suite *GroupsIntgSuite) TestGetAllTeams() {
...
 
-	for _, team := range groups {
-		assert.True(t, api.FetchOnlyTeams(ctx, team), "must not return non groups groups")
+	var isTeam bool
+
+	if len(groups) > len(teams) {
+		isTeam = true
 	}
+
+	assert.True(t, isTeam, "must not return non groups ")
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/groups_test.go]:
```suggestion
	assert.True(t, len(groups) > len(teams), "must only return teams")
```

File: src/pkg/services/m365/api/groups.go [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	if err := iter.Iterate(ctx, iterator); err != nil {
+		return nil, graph.Wrap(ctx, err, "iterating all groups")
+	}
+
+	return groups, el.Failure()
+}
+
+func FetchOnlyTeams(ctx context.Context, groups []models.Groupable) []models.Groupable {
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/groups.go:116]:
```suggestion
func OnlyTeams(ctx context.Context, groups []models.Groupable) []models.Groupable {
```

File: src/pkg/services/m365/api/groups.go (hunk 21 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/groups.go:121]: I am little confused here. We want to continue is there is some value in `g.GetAdditionalData()[ResourceProvisioningOptions]` ?

File: src/pkg/services/m365/api/groups.go (hunk 22 above)
Comment (neha-Gupta1) [reply src/pkg/services/m365/api/groups.go:132]: its used only here and not at multiple places.
